babel.config.mjs not working

**Current behavior**
On node@14.8.0, I have 2 equivalent simple files: babel.config.cjs and babel.config.mjs (except for the exporting keyword)...

```js
const presets = [
  [
    "@babel/env",
    {
      targets: {
        edge: "10",
        firefox: "50",
        chrome: "60",
        safari: "11.1",
      },
      useBuiltIns: "usage",
      corejs: 2,
    }
  ]
];

module.exports = { presets };
```

commenting the mjs, everything works perfect doing npx babel src/index.js, while commenting the cjs one (enabling esm support), I get 
"Error: /Users/kotaro/Projects/ro/test/babel.config.mjs: Error while loading config - Cannot find module 'file:///Users/kotaro/Projects/ro/test/babel.config.mjs'
Require stack:
- /Users/kotaro/Projects/ro/test/node_modules/@babel/core/lib/config/files/import.js
- /Users/kotaro/Projects/ro/test/node_modules/@babel/core/lib/config/files/module-types.js
- /Users/kotaro/Projects/ro/test/node_modules/@babel/core/lib/config/files/configuration.js
- /Users/kotaro/Projects/ro/test/node_modules/@babel/core/lib/config/files/index.js
- /Users/kotaro/Projects/ro/test/node_modules/@babel/core/lib/index.js
- /Users/kotaro/Projects/ro/test/node_modules/@babel/cli/lib/babel/options.js
- /Users/kotaro/Projects/ro/test/node_modules/@babel/cli/lib/babel/index.js
- /Users/kotaro/Projects/ro/test/node_modules/@babel/cli/bin/babel.js"

Tracking the sources, it seems you guys are transpiling import into a require at https://github.com/babel/babel/blob/main/packages/babel-core/src/config/files/import.js:

```js
function import_(filepath) {
  console.log('000000', filepath);
  return Promise.resolve(`${filepath}`).then(s => _interopRequireWildcard(require(s)));
}
```
so that filepath is an absolute URL: { href: 'file:///Users/kotaro/Projects/ro/test/babel.config.mjs', ...} which seems to fail loading (yes, file is there).

Setting type: "module" in package json provokes same error

**Expected behavior**
babel should handle equivalently cjs and mjs if node version supports it

**Environment**
node@14.8.0

```
- Babel version(s): 7.11.1
- Node/npm version: [e.g. Node 14.8.0/npm 6.14.7]
- OS: OSX 10.13.4
- Monorepo: no
- How you are using Babel: cli ATM

**Possible Solution**


**Additional context**
Add any other context about the problem here. Or a screenshot if applicable

--- a/Gulpfile.js
+++ b/Gulpfile.js
@@ -69,7 +69,14 @@ function buildBabel(exclude, sourcesGlob = defaultSourcesGlob) {
     .pipe(errorsLogger())
     .pipe(newer({ dest: base, map: swapSrcWithLib }))
     .pipe(compilationLogger())
-    .pipe(babel())
+    .pipe(
+      babel({
+        caller: {
+          // We have wrapped packages/babel-core/src/config/files/configuration.js with feature detection
+          supportsDynamicImport: true,
+        },
+      })
+    )
     .pipe(
       // Passing 'file.relative' because newer() above uses a relative
       // path and this keeps it consistent.

--- a/packages/babel-core/test/config-chain.js
+++ b/packages/babel-core/test/config-chain.js
@@ -1,6 +1,8 @@
+import cp from "child_process";
 import fs from "fs";
 import os from "os";
 import path from "path";
+import util from "util";
 import escapeRegExp from "lodash/escapeRegExp";
 import * as babel from "../lib";
 
@@ -51,6 +53,31 @@ function loadOptionsAsync(opts) {
   return babel.loadOptionsAsync({ cwd: __dirname, ...opts });
 }
 
+async function loadOptionsAsyncInSpawedProcess({ filename, cwd }) {
+  // !!!! hack is coming !!!!
+  // remove this section when https://github.com/facebook/jest/issues/9430 is resolved
+  // We don't check process.versions.node here, it will fail if node does not support esm
+  // please publish Babel on a modernized node :)
+  const { stdout, stderr } = await util.promisify(cp.execFile)(
+    require.resolve("./fixtures/babel-load-options-async.mjs"),
+    // pass `cwd` as params as `process.cwd()` will normalize `cwd` on macOS
+    [filename, cwd],
+    {
+      cwd,
+      env: process.env,
+    },
+  );
+  if (stderr) {
+    throw new Error(
+      "error is thrown in babel-load-options-async.mjs: stdout\n" +
+        stdout +
+        "\nstderr:\n" +
+        stderr,
+    );
+  }
+  return JSON.parse(stdout);
+}
+
 function pairs(items) {
   const pairs = [];
   for (let i = 0; i < items.length - 1; i++) {
@@ -61,11 +88,10 @@ function pairs(items) {
   return pairs;
 }
 
-async function getTemp(name) {
+async function getTemp(name, fixtureFolder = "config-files-templates") {
   const cwd = await pfs.mkdtemp(os.tmpdir() + path.sep + name);
   const tmp = name => path.join(cwd, name);
-  const config = name =>
-    pfs.copyFile(fixture("config-files-templates", name), tmp(name));
+  const config = name => pfs.copyFile(fixture(fixtureFolder, name), tmp(name));
   return { cwd, tmp, config };
 }
 
@@ -1034,20 +1060,20 @@ describe("buildConfigChain", function () {
         );
       });
 
-      test.each([
-        "babel.config.json",
-        "babel.config.js",
-        "babel.config.cjs",
-        "babel.config.mjs",
-      ])("should load %s asynchronously", async name => {
+      test.each(
+        [
+          "babel.config.json",
+          "babel.config.js",
+          "babel.config.cjs",
+          // We can't transpile import() while publishing, and it isn't supported
+          // by jest.
+          process.env.IS_PUBLISH ? "" : "babel.config.mjs",
+        ].filter(Boolean),
+      )("should load %s asynchronously", async name => {
         const { cwd, tmp, config } = await getTemp(
           `babel-test-load-config-async-${name}`,
         );
         const filename = tmp("src.js");
-
-        // We can't transpile import() while publishing, and it isn't supported
-        // by jest.
-        if (process.env.IS_PUBLISH && name === "babel.config.mjs") return;
 
         await config(name);
 
@@ -1087,6 +1113,29 @@ describe("buildConfigChain", function () {
           loadOptionsAsync({ filename: tmp("src.js"), cwd }),
         ).rejects.toThrow(/Multiple configuration files found/);
       });
+
+      if (process.env.IS_PUBLISH) {
+        test.each(["babel.config.mjs", ".babelrc.mjs"])(
+          "should load %s asynchronously",
+          async name => {
+            const { cwd, tmp, config } = await getTemp(
+              `babel-test-load-config-async-prepublish-${name}`,
+              "config-files-templates-prepublish",
+            );
+            const filename = tmp("src.js");
+            await config(name);
+            expect(
+              await loadOptionsAsyncInSpawedProcess({ filename, cwd }),
+            ).toEqual({
+              ...getDefaults(),
+              filename,
+              cwd,
+              root: cwd,
+              comments: true,
+            });
+          },
+        );
+      }
     });
 
     describe("relative", () => {
@@ -1126,21 +1175,21 @@ describe("buildConfigChain", function () {
         );
       });
 
-      test.each([
-        "package.json",
-        ".babelrc",
-        ".babelrc.js",
-        ".babelrc.cjs",
-        ".babelrc.mjs",
-      ])("should load %s asynchronously", async name => {
+      test.each(
+        [
+          "package.json",
+          ".babelrc",
+          ".babelrc.js",
+          ".babelrc.cjs",
+          // We can't transpile import() while publishing, and it isn't supported
+          // by jest.
+          process.env.IS_PUBLISH ? "" : "babel.config.mjs",
+        ].filter(Boolean),
+      )("should load %s asynchronously", async name => {
         const { cwd, tmp, config } = await getTemp(
           `babel-test-load-config-${name}`,
         );
         const filename = tmp("src.js");
-
-        // We can't transpile import() while publishing, and it isn't supported
-        // by jest.
-        if (process.env.IS_PUBLISH && name === ".babelrc.mjs") return;
 
         await config(name);
 

--- a/packages/babel-core/test/config-chain.js
+++ b/packages/babel-core/test/config-chain.js
@@ -55,7 +55,7 @@ function loadOptionsAsync(opts) {
 
 async function loadOptionsAsyncInSpawedProcess({ filename, cwd }) {
   // !!!! hack is coming !!!!
-  // remove this section when https://github.com/facebook/jest/issues/9430 is resolved
+  // todo(Babel 8): remove this section when https://github.com/facebook/jest/issues/9430 is resolved
   // We don't check process.versions.node here, it will fail if node does not support esm
   // please publish Babel on a modernized node :)
   const { stdout, stderr } = await util.promisify(cp.execFile)(

--- a/packages/babel-core/test/fixtures/babel-load-options-async.mjs
+++ b/packages/babel-core/test/fixtures/babel-load-options-async.mjs
@@ -1,0 +1,13 @@
+#!/usr/bin/env node
+
+// Usage:
+// babel-load-options-async.js [filename]
+import babel from "@babel/core";
+
+const [, , filename, cwd] = process.argv;
+
+(async () => {
+  process.stdout.write(
+    JSON.stringify(await babel.loadOptionsAsync({ filename, cwd }))
+  );
+})();

--- a/packages/babel-core/test/fixtures/config/config-files-templates-prepublish/.babelrc.mjs
+++ b/packages/babel-core/test/fixtures/config/config-files-templates-prepublish/.babelrc.mjs
@@ -1,0 +1,3 @@
+export default {
+  comments: true,
+};

--- a/packages/babel-core/test/fixtures/config/config-files-templates-prepublish/babel.config.mjs
+++ b/packages/babel-core/test/fixtures/config/config-files-templates-prepublish/babel.config.mjs
@@ -1,0 +1,3 @@
+export default {
+  comments: true,
+};

